storage-proofs/post/src/fallback/vanilla.rs

```diff
@@ -332,9 +342,33 @@ impl<'a, Tree: 'a + MerkleTreeTrait> ProofScheme<'a> for FallbackPoSt<'a, Tree>
                             challenge_index,
                         )?;
 
-                        tree.gen_cached_proof(challenged_leaf_start as usize, Some(rows_to_discard))
+                        let proof = tree.gen_cached_proof(
+                            challenged_leaf_start as usize,
+                            Some(rows_to_discard),
+                        );
+                        let valid = if let Ok(proof) = &proof {
+                            proof.validate(challenged_leaf_start as usize)
+                                && proof.root() == priv_sector.comm_r_last
+                        } else {
+                            false
+                        };
+                        if valid {
+                            Ok(ProofOrFault::Proof(proof))
+                        } else {
+                            Ok(ProofOrFault::Fault(sector_id))
+                        }
                     })
-                    .collect::<Result<Vec<_>>>()?;
+                    .collect::<Result<Vec<_>>>()?
+                {
+                    match proof_or_fault {
+                        ProofOrFault::Proof(proof) => {
+                            inclusion_proofs.push(proof?);
+                        }
+                        ProofOrFault::Fault(sector_id) => {
+                            faulty_sectors.insert(sector_id);
+                        }
+                    }
+                }
```

Comment:
That might not work well with the `into_par_iter()` (I haven't tried), but is the whole for loop and the `ProofOrFault` enum needed at all? Couldn't you just push things directly into the vectors?

```rust
match &proof {
    Ok(proof) => {
        if proof.validate(challenged_leaf_start as usize)
            && proof.root() == priv_sector.comm_r_last {
            inclusion_proofs.push(proof?);
        } else {
            faulty_sectors.insert(sector_id);
        }
    }
    Err(_) => faulty_sectors.insert(sector_id);
}
```